### PR TITLE
W8 - D3 - HungTran - Canvas

### PR DESCRIPTION
**Pull request by at-hungtran2**
#49

### CODE REVIEW

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        canvas.translate(mTranslateX, 0);
+        drawLine(canvas, top);
+        drawChart(canvas, top, space, distance);
+        canvas.restore();
+        canvas.scale(mScaleFactor, mScaleFactor, mDetector.getFocusX(), mDetector.getFocusY());
+        drawRectBackground(canvas, top, unit);
+        drawTextSaleandExpense(canvas, top);
+    }
+
+    @SuppressWarnings("deprecation")
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: delete this and fix warning.

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    private void initPaint() {
+        createListMoney();
+        WindowManager w = ((Activity) getContext()).getWindowManager();
+        float element = (w.getDefaultDisplay().getWidth() - 110) / (13.0f * 6);
+        mTotalWidth = 80 + 7 * element * mMoney.size() + 6 * element * (mMoney.size() - 1);
+        setFadingEdgeLength(0);
+        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
+        mPaint.setColor(getResources().getColor(R.color.colorGray));
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: use `ContextCompat.getColor()`

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        mPaint.setStrokeWidth(1);
+    }
+
+    private void createListMoney() {
+        mMoney.add(new Money("Jan", 75000, 10000));
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: try with large value `750000`?

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        mPaint.setStyle(Paint.Style.FILL);
+        mPaint.setColor(getResources().getColor(R.color.colorGray));
+        while (n <= 7) {
+            canvas.drawText(getResources().getString(R.string.dola) + n * 20000, 10, top - n * top / 12 + 10, mPaint);
+            n++;
+        }
+    }
+
+    private void drawRectBackground(Canvas canvas, float top, float unit) {
+        mPaint.setColor(getResources().getColor(R.color.colorWhite));
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: color need put in `attrs.xml`

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        mPaint.setStyle(Paint.Style.FILL);
+        canvas.drawRect(getWidth() - 5, top - 150 - unit * maxChartValue(mMoney), getWidth(), top + 50, mPaint);
+        canvas.drawRect(0, top - 150 - unit * maxChartValue(mMoney), 120, top + 50, mPaint);
+        mPaint.setColor(getResources().getColor(R.color.colorGray));
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: define color in `attrs.xml` and set to view.

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    }
+
+    private void drawSetTitle(Canvas canvas, float top, float unit) {
+        canvas.drawText(getResources().getString(R.string.title_chart), (getWidth() - 500) / 2.0f, top - 80 - unit * maxChartValue(mMoney), mPaint);
+        mPaint.setStyle(Paint.Style.FILL);
+        mPaint.setTextSize(getContext().getResources().getDimension(R.dimen.text_size_name));
+    }
+
+    private void drawRectSaleandExpense(Canvas canvas, float top, float unit) {
+
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: delete this line

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+                canvas.drawLine(180, line, getWidth() - 30 - 3 * element, line, mPaint);
+                line -= space;
+                n += distance;
+            }
+        }
+        mPaint.setColor(getResources().getColor(R.color.colorGray));
+        mPaint.setStyle(Paint.Style.FILL);
+        for (Money money : mMoney) {
+            mPaint.setColor(getResources().getColor(R.color.colorGray));
+            mPaint.setTextSize(25);
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: `textSize` put in `dimens.xml`

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+        public boolean onScale(ScaleGestureDetector detector) {
+            mScaleFactor *= detector.getScaleFactor();
+            mScaleFactor = Math.max(MIN_ZOOM, Math.min(mScaleFactor, MAX_ZOOM));
+            if (mScaleFactor > 1f) {
+                invalidate();
+            }
+            return true;
+        }
+    }
+}
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: enter

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        mPaint.setStyle(Paint.Style.FILL);
+        mPaint.setTextSize(getContext().getResources().getDimension(R.dimen.text_size_name));
+    }
+
+    private void drawRectSaleandExpense(Canvas canvas, float top, float unit) {
+
+        canvas.drawText(getResources().getString(R.string.sale), 800, top + 70, mPaint);
+        canvas.drawText(getResources().getString(R.string.expense), 1000, top + 70, mPaint);
+        mPaint.setColor(getResources().getColor(R.color.blue));
+        canvas.drawRect(770, top + 50, 790, top + 70, mPaint);
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: `width` of column get from `attrs.xml`

Comment (at-trungnguyen) [reply app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: you fix width of column?

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+    private void drawRectBackground(Canvas canvas, float top, float unit) {
+        mPaint.setColor(getResources().getColor(R.color.colorWhite));
+        mPaint.setStyle(Paint.Style.FILL);
+        canvas.drawRect(getWidth() - 5, top - 150 - unit * maxChartValue(mMoney), getWidth(), top + 50, mPaint);
+        canvas.drawRect(0, top - 150 - unit * maxChartValue(mMoney), 120, top + 50, mPaint);
+        mPaint.setColor(getResources().getColor(R.color.colorGray));
+        mPaint.setStyle(Paint.Style.STROKE);
+        canvas.drawRect(5, top - 150 - unit * maxChartValue(mMoney), getWidth() - 5, top + 100, mPaint);
+
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: delete this line

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+            }
+        }
+        mPaint.setColor(getResources().getColor(R.color.colorGray));
+        mPaint.setStyle(Paint.Style.FILL);
+        for (Money money : mMoney) {
+            mPaint.setColor(getResources().getColor(R.color.colorGray));
+            mPaint.setTextSize(25);
+            canvas.drawText(String.valueOf(money.getMonth()), left + 60, top + 25, mPaint);
+            mPaint.setColor(getResources().getColor(R.color.blue));
+            canvas.drawRect(left, top, left += element * 3, top - money.getSale() * top / 12 / 20000, mPaint);
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: why `20000`, you should calculate depend on value you input.

File: app/src/at-hungtran/res/layout/activity_canvas.xml [hunk 12]
```diff
@@ -0,0 +1,12 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/frameLayoutChart"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context="asiantech.internship.summer.canvas.CanvasActivity">
+
+    <asiantech.internship.summer.canvas.GraphicViewSale
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/res/layout/activity_canvas.xml]: if layout have `1 view` you can put `<asiantech.internship.summer.canvas.GraphicViewSale` as parent, no need `LinearLayout` here.

File: app/src/at-hungtran/res/layout/activity_canvas.xml [hunk 13]
```diff
@@ -0,0 +1,12 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/frameLayoutChart"
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/res/layout/activity_canvas.xml]: this `id` use for what?

File: app/src/at-hungtran/res/values/colors.xml [hunk 14]
```diff
@@ -7,4 +7,6 @@
     <color name="colorWhite">#FFFFFF</color>
     <color name="colorGrey">#aea5a5</color>
     <color name="colorBlack">#020202</color>
+    <color name="colorOrange">#ff4901</color>
```
Comment (at-trungnguyen) [inline app/src/at-hungtran/res/values/colors.xml]: color is a param that you can input in layout, so it need to define in `attrs.xml`

File: app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java (hunk 9 above)
Comment (at-hungtran2) [reply app/src/at-hungtran/java/asiantech/internship/summer/canvas/GraphicViewSale.java]: A @at-trungnguyen em k hieu comment nay lam'

**Pull request merged**